File: leetcode/best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
  int maxProfit(vector<int> &prices) {
    int m = 0, low = prices[0], l = prices.size(), i = 0;

    while (i < l - 1) {
      while (i + 1 < l && prices[i + 1] <= prices[i]) {
        i++;
      }
      low = prices[i] < low ? min(low, prices[i]);

      while (i + 1 < l && prices[i + 1] >= prices[i]) {
        i++;
      }
      m = max(m, prices[i] - low);
    }

    return m;
  }
};

File: leetcode/best-time-to-buy-and-sell-stock.go
func maxProfit(prices []int) int {
	l, max, low, i := len(prices)-1, 0, prices[0], 0

	for i < l {
		for i < l && prices[i+1] <= prices[i] {
			i++
		}

		if prices[i] < low {
			low = prices[i]
		}

		for i < l && prices[i+1] >= prices[i] {
			i++
		}

		if prices[i]-low > max {
			max = prices[i] - low
		}
	}
	return max
}
